src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementEventNormalizationService.cs

```diff
@@ -53,30 +54,31 @@ private static ISourceBlock<EventData> StartProducer(IEnumerable<EventData> data
         {
             var producer = new BufferBlock<EventData>(new DataflowBlockOptions { BoundedCapacity = DataflowBlockOptions.Unbounded });
 
-            _ = Task.Run(async () =>
-              {
-                  foreach (var evt in data)
-                  {
-                      while (!await producer.SendAsync(evt))
-                      {
-                          await Task.Yield();
-                      }
-                  }
-
-                  producer.Complete();
-              });
+            _ = Task.Run(async () =>
+            {
+                foreach (var evt in data)
+                {
+                    while (!await producer.SendAsync(evt))
+                    {
+                        await Task.Yield();
+                    }
+                }
+
+                producer.Complete();
+            });
 
             return producer;
         }
 
-        private async Task StartConsumer(ISourceBlock<EventData> producer, IAsyncCollector<IMeasurement> collector, Func<Exception, EventData, Task<bool>> errorConsumer)
+        private async Task StartConsumer(ISourceBlock<EventData> producer, IEnumerableAsyncCollector<IMeasurement> collector, Func<Exception, EventData, Task<bool>> errorConsumer)
         {
             // Collect non operation canceled exceptions as they occur to ensure the entire data stream is processed
             var exceptions = new ConcurrentBag<Exception>();
             var cts = new CancellationTokenSource();
-            var consumer = new ActionBlock<EventData>(
+            var transformingConsumer = new TransformManyBlock<EventData, IMeasurement>(

```

Comment:
It isn't immediately clear to me the usage of the TransformManyBlock here.  Can you describe conceptually what you are doing here please?

Reply:
It's a TPL block that can take an input X and transforms it into an enumerable of Y. Here, I receive a single `EventData` object and produce an enumerable of `IMeasurement` from it. The collection of `IMeasurement` then gets output into the data processing pipeline to later get batched and output to EventHub.

Reply:
Okay the one part that isn't clear to me is at what point does the transform block submit its output to the action block that sends the data to EventHub.  Is everything thing from the producer exhausted and buffered in the transform block and everything flushed at the end once the source is exhausted or is it flushed for every input event.

For example if I have 2 events to process, the first one emits 3 measurements  and the second emits 2 measurements.  Does final action block get called twice (once with 3 measurements and another with two) or does it get called once with 5?

If it is the former then this batching really only helps if we are doing projections.  If have I have two distinct messages and each nly output 1 measurement we would still have 2 different requests to the destination event hub (one each).  This is how it appears to work based on my current reading of the code and research.  Does that match your understanding?

If so I am okay with the change, seems relatively safe.  We potentially could make additional modifications to batch even more but it might add additional latency which may not be desirable.

Reply:
Based on my understanding, in your example it would get called once with 5 measurements.

The BatchBlock is invoked when either the amount of message batched equals `_asyncCollectorBatchSize` or the TransformBlock completes. In your example, the 5 message wouldn't hit the batch size but once the producer finishes the completion will propogate up the pipeline, causing the TransformBlock to finish, which forces the BatchBlock to finish. At this point it will flush out its batched 5 messages to the ActionBlock so that the 5 get written out as batch to EventHub.

Reply:
FYI, the full TPL pipeline is:
producer -> TransformBlock -> BatchBlock -> ActionBlock. Its in the BatchBlock that message get batched before they are sent to the ActionBlock (i.e. EventHub)

Reply:
Gothca I missed the batch block sorry.  So my understanding would be correct with out hte batch block